apps/prairielearn/src/lib/assets.ts

```diff
@@ -134,13 +134,20 @@ function assertAssetsPrefix(): string {
   return assetsPrefix;
 }
 
+let initialized = false;
+
 /**
  * Computes the hashes of directories from which we serve cacheable assets.
  * Should be run at server startup before any responses are served.
  *
  * Also initializes the assets compiler.
  */
 export async function init() {
```

Comment:
is it assumed that `init` methods are only for testing?

Reply:
If not, we should follow a similar pattern to #12020

Reply:
If they are, then this PR should fixup #12020 to always enable it -- it would be very confusing to have some inits for testing only, and some that are not.

Reply:
No, they're never only for testing:

https://github.com/PrairieLearn/PrairieLearn/blob/9b93f33429f0addb648a6620a9d51a3d58dce78d/apps/prairielearn/src/server.ts#L2361

Why should this be made conditional? This behavior change doesn't matter for normal runtime (production, etc.) because it's guaranteed that this is only called once per processes.

I don't know what you mean by "always enable it" - always enable what?

Reply:
I see. The comment had confused me about the fact that  for normal runtime "it's guaranteed that this is only called once" per process. I would imagine that some error handling/recovery could break that assumption.

Perhaps you can add a comment to `helperServer.ts` stating the "only called once" assumption.

Reply:
It's specifically _not_ only called once per process in tests; that's why this change nets us any improvements.

Can you propose a specific comment and line to place it that would have helped you understand what's going on here?

Reply:
```
// This hook may be called repeatedly for test setup and cleanup, but is guaranteed to only be called once during normal execution (initialization code that accumulates state only affects tests)
```

In `apps/prairielearn/src/tests/helperServer.ts`, line 35.

Reply:
> It's specifically not only called once per process in tests; that's why this change nets us any improvements.

It was the assumption that in production settings, this function is guaranteed to only be called once that confused me.

Reply:
This pattern of code (obviously not how I would write it)
```ts

try {
  init()
} catch (fail) {
  if (fail == transient) {
    sleep(5)
    init()
  }
}
```

Reply:
Wait, I'm more confused now. Why line 35? These test utilities (`before` and such) have nothing to do with the real server.

Reply:
Sorry, that's a bad spot. The `init` call in `assets` sounds good.

Reply:
Candidly we'd never write test code like that 😅 if there are transient failures, we'd identify a root cause and fix them, not blindly retry things. Other pieces of the this overall process aren't idempotent.

Reply:
I added a different comment in what I felt to be a more appropriate location in 9f6b268.